_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+OKTA_OAUTH2_CLIENT_ID=
+OKTA_OAUTH2_CLIENT_SECRET=
+OKTA_OAUTH2_AUDIENCE=https://<your-auth0-domain>/api/v2/
+```
+
+At startup, these properties will be read using [spring-dotenv](https://github.com/paulschwarz/spring-dotenv).
+
+Run `./gradlew bootRun` to start the API gateway, or use your IDE to run it.
+
+Copy `car-service/.env.example` to `.env` and update its values.
+
+```dotenv
+OKTA_OAUTH2_ISSUER=https://<your-auth0-domain>/
+OKTA_OAUTH2_AUDIENCE=https://<your-auth0-domain>/api/v2/
+```
+
+Start it with `./gradlew bootRun` and open `http://localhost:8080` in your favorite browser. You'll be redirected to Auth0 to log in.
+
+{% img blog/spring-boot-microservices/auth0-login.png alt:"Auth0 Login" %}{: .center-image }
+
+After authenticating, you'll see your name in lights! ✨
+
+{% img blog/spring-boot-microservices/name-in-lights.png alt:"Your name in lights" %}{: .center-image }
+
+You can navigate to the following URLs in your browser for different results:
+
+- `http://localhost:8080/print-token`: prints access token to the console
+- `http://localhost:8080/cool-cars`: returns a list of cool cars
+- `http://localhost:8080/home`: proxies request to the car service and prints JWT claims to the console
+
+You can see the access token's contents by copying/pasting it into [jwt.io](https://jwt.io). You can also access the car service directly using it.
+
+```shell
+TOKEN=<access-token>
+http :8090/cars Authorization:"Bearer $TOKEN"
+```
+
+{% img blog/spring-boot-microservices/car-service-with-token.png alt:"Car Service with HTTPie and access token" %}{: .center-image }
+
+Pretty cool, eh? 😎
+
+## My Developer Story with Spring Boot and Spring Cloud
+
+A few years ago, I created an [example similar to this one](http://developer.okta.com/blog/2019/05/22/java-microservices-spring-boot-spring-cloud) with Spring Boot 2.2. It used Feign for remote connectivity, Zuul for routing, Hystrix for failover, and Spring Security for OAuth. The September 2023 version of Spring Cloud has [Spring Cloud OpenFeign](https://spring.io/projects/spring-cloud-openfeign) for remote connectivity, [Spring Cloud Gateway](https://spring.io/projects/spring-cloud-gateway) for routing, and [Resilience4j](https://resilience4j.readme.io/) for fault tolerance.
+
+Okta also now has an [Okta Spring Boot starter](https://github.com/okta/okta-spring-boot). I didn't use it in my first experiment, but I'm a big fan of it after the last few years! It dramatically simplifies configuration and makes securing your apps with OAuth 2.0 and OIDC easy. It's a thin wrapper around Spring Security's resource server, OAuth client, and OIDC features. Not only that, but it works with Okta Workforce Identity, Okta Customer Identity (aka Auth0), and even Keycloak.
+
+I created all of these applications using [start.spring.io](https://start.spring.io)'s REST API and [HTTPie](https://httpie.org).
+
+```shell
+https start.spring.io/starter.zip bootVersion==3.1.4 \
```

Comment:
bootVersion is not required IMO. I removed them in the workshop

Reply:
It isn't, but things might stop working with Spring Boot 3.2 because they sometimes break backward compatibility. I requested start.spring.io make things easier: https://github.com/spring-io/start.spring.io/issues/1303 I'd appreciate an upvote!